pkg/helm/watcher/controller/helm_watcher.go

```diff
@@ -140,3 +159,82 @@ func (r *HelmWatcherReconciler) reconcileDelete(ctx context.Context, repository
 	// Stop reconciliation as the object is being deleted
 	return ctrl.Result{}, nil
 }
+
+// sendEvent emits a Kubernetes event and forwards the event to notification controller if configured.
+func (r *HelmWatcherReconciler) sendEvent(log logr.Logger, hr *sourcev1.HelmRepository, revision, severity, profileName, version string) {
+	if r.ExternalEventRecorder == nil {
+		return
+	}
+
+	objRef, err := reference.GetReference(r.Scheme, hr)
+	if err != nil {
+		log.Error(err, "unable to get reference")
+		return
+	}
+
+	var meta map[string]string
+	if revision != "" {
+		meta = map[string]string{"revision": revision}
+	}
+
+	if err := r.ExternalEventRecorder.EventInfof(*objRef, meta, severity, "New version available for profile %s with version %s", profileName, version); err != nil {
+		log.Error(err, "unable to send event")
+		return
+	}
+}
+
+// checkForNewVersion uses existing data to determine if there are newer versions in the incoming data
+// compared to what's already stored in the cache. It returns the LATEST version which is greater than
+// the last version that was stored.
+func (r *HelmWatcherReconciler) checkForNewVersion(ctx context.Context, chart *pb.Profile) (string, error) {
+	versions, err := r.Cache.ListAvailableVersionsForProfile(ctx, chart.GetHelmRepository().GetNamespace(), chart.GetHelmRepository().GetName(), chart.Name)
+	if err != nil {
+		return "", err
+	}
+
+	newVersions, err := r.convertStringListToSemanticVersionList(chart.AvailableVersions)
+	if err != nil {
+		return "", err
+	}
+
+	oldVersions, err := r.convertStringListToSemanticVersionList(versions)
+	if err != nil {
+		return "", err
+	}
+
+	sort.SliceStable(newVersions, func(i, j int) bool {
+		return newVersions[i].GreaterThan(newVersions[j])
+	})
+
+	sort.SliceStable(oldVersions, func(i, j int) bool {
+		return oldVersions[i].GreaterThan(oldVersions[j])
+	})
+
+	// If there are no old versions stored, it's likely that the profile didn't exist before. So we don't notify.
+	// Same in case there are no new versions ( which is unlikely to happen, but we ward against it nevertheless ).
+	if len(oldVersions) == 0 || len(newVersions) == 0 {
+		return "", nil
+	}
+
+	// Notify in case the latest new version is greater than the latest old version.
+	if newVersions[0].GreaterThan(oldVersions[0]) {
+		return newVersions[0].String(), nil
+	}
+
+	return "", nil
+}
+
+func (r *HelmWatcherReconciler) convertStringListToSemanticVersionList(versions []string) ([]*semver.Version, error) {
+	var result []*semver.Version
+
+	for _, v := range versions {
+		ver, err := semver.NewVersion(v)
+		if err != nil {
+			return nil, err
+		}
```

Comment:
felt this pain in eksctl too but it feels so frustrating to blow up (error) if 1 out of 100 versions can't be parsed. Perhaps worth just skipping the version and logging a warning instead?

Reply:
Sometimes I'd say yes, but in this case, there are a number of things upstream that should prevent this, so if a chart has evaded these checks, I'd say it's reasonable to blow up :-)

Reply:
Fair point! Sounds good